lib/player.js

```diff
@@ -2077,30 +2037,42 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
           });
     }
 
-    // Give the user a chance to change things and filter again, so they can
-    // influence what the initial variant is.
+    // The event must be fired after we filter by restrictions but before the
+    // active stream is picked to allow those listening for the "streaming"
+    // event to make changes before streaming starts.
     this.dispatchEvent(
         this.makeEvent_(shaka.util.FakeEvent.EventName.Streaming));
+
+    // Pick the initial streams to play.
+    // Unless the user has already picked a variant, anyway, by calling
+    // selectVariantTrack before this loading stage.
+    let initialVariant = null;
+    const activeVariantTrack = this.getVariantTracks().find((t) => t.active);
     if (!activeVariantTrack) {
-      const oldVariant = initialVariant;
       initialVariant = this.chooseVariant_();
       goog.asserts.assert(initialVariant, 'Must choose an initial variant!');
-      if (oldVariant != initialVariant) {
-        // The user did something in the streaming event that changes the
-        // variant we would have picked, so lazy-load that variant too.
-        await shaka.Player.loadVariant_(initialVariant);
-        // Also unload the segment index of the previous initial variant, so as
-        // to not waste memory on it.
-        for (const stream of [oldVariant.video, oldVariant.audio]) {
-          if (stream && stream.closeSegmentIndex) {
-            stream.closeSegmentIndex();
-          }
+
+      // Lazy-load it, so we will have enough info to make the playhead.
```

Comment:
It looks to me like this should be done outside the conditional.  If someone picks streams in a `streaming` event handler, then activeVariantTrack will be non-null.  Wouldn't you still need those track's SegmentIndexes to be available for playhead?

Reply:
Good point. Done.